lib/awful/spawn.lua

```diff
@@ -435,6 +470,149 @@ function spawn.read_lines(input_stream, line_callback, done_callback, close)
+-- This function tries its best to preserve the state across `awesome.restart()`.
+--
+-- By default, when no `uid` is specified, this function will generate one by
+-- hashing the command and its rules. If you have multiple instance of the same
+-- command and rules, you need to specify an UID or only the first one will be
+-- executed.
+--
+-- The `rules` are standard `awful.rules`.
+--
+-- This function depends on the startup notification protocol to be correctly
+-- implemented by the command. See `client.startup_id` for more information.
+-- Note that this also wont work with shell or terminal commands.
+--
+-- @tparam string|table cmd The command.
+-- @tparam table rules The properties that need to be applied to the client.
+-- @tparam[opt] string uid A string to identify the client so it isn't executed
+--  multiple time.
+-- @tparam[opt] function matcher A matching function to find the instance
+--  among running clients.
+-- @tparam[opt] function callback A callback function when the client is created.
+-- @see awful.rules
+function spawn.once(cmd, rules, uid, matcher, callback)
+    local hash = uid or hash_command(cmd, rules)
+
+    local status = register_common(hash, rules, callback)
+
+    local function do_it()
+        if not status.exec and not is_running(hash, matcher) then
+            run_once_common(hash, cmd)
+        end
+    end
+
+    -- The clients are not yet managed during the first execution, so it will
+    -- miss existing instances.
+    if awesome.startup then
+        gtimer.delayed_call(do_it)
+    else
+        do_it()
+    end
+end
+
+--- Spawn a command if an instance is not already running.
+--
+-- This is like `awful.spawn.once`, but will spawn new instances if the previous
+-- has finished.
+--
+-- The `rules` are standard `awful.rules`.
+--
+-- This function depends on the startup notification protocol to be correctly
+-- implemented by the command. See `client.startup_id` for more information.
+-- Note that this also wont work with shell or terminal commands.
+--
+-- @tparam string|table cmd The command.
+-- @tparam table rules The properties that need to be applied to the client.
+-- @tparam[opt] string uid A string to identify the client so it isn't executed
+--  multiple time.
+-- @tparam[opt] function matcher A matching function to find the instance
+--  among running clients.
+-- @tparam[opt] function callback A callback function when the client is created.
+-- @see awful.rules
+function spawn.single_instance(cmd, rules, uid, matcher, callback)
+    local hash = uid or hash_command(cmd, rules)
+
+    register_common(hash, rules, callback)
+
+    local function do_it()
+        if not is_running(hash, matcher) then
+            return run_once_common(hash, cmd)
+        end
+    end
+
+    -- The clients are not yet managed during the first execution, so it will
+    -- miss existing instances.
+    if awesome.startup then
+        gtimer.delayed_call(do_it)
+    else
+        do_it()
+    end
+end
```

Comment:
After taking a closer look: Both spawning functions are basically the same. The only difference is in checking `status.exec`. It should be possible to merge them so that both functions just call the same helper function with an extra boolean parameter.